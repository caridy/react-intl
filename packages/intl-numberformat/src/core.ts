import {
  defineProperty,
  invariant,
  RawNumberLocaleData,
  SupportedLocales,
  unpackData,
  InitializeNumberFormat,
  FormatNumericToParts,
  NumberFormatOptions,
  ToNumber,
  CanonicalizeLocaleList,
} from '@formatjs/ecma402-abstract';
import * as currencyDigitsData from './data/currency-digits.json';
import {names as numberingSystemNames} from './data/numbering-systems.json';

// eslint-disable-next-line import/no-cycle
import getInternalSlots from './get_internal_slots';
import {
  NumberFormatConstructor,
  NumberFormat as NumberFormatType,
} from './types';

// Merge declaration with the constructor defined below.
export type NumberFormat = NumberFormatType;

const RESOLVED_OPTIONS_KEYS = [
  'locale',
  'numberingSystem',
  'style',
  'currency',
  'currencyDisplay',
  'currencySign',
  'unit',
  'unitDisplay',
  'minimumIntegerDigits',
  'minimumFractionDigits',
  'maximumFractionDigits',
  'minimumSignificantDigits',
  'maximumSignificantDigits',
  'useGrouping',
  'notation',
  'compactDisplay',
  'signDisplay',
] as const;

/**
 * https://tc39.es/ecma402/#sec-intl-numberformat-constructor
 */
export const NumberFormat = function (
  this: NumberFormat,
  locales?: string | string[],
  options?: NumberFormatOptions
) {
  // Cannot use `new.target` bc of IE11 & TS transpiles it to something else
  if (!this || !(this instanceof NumberFormat)) {
    return new NumberFormat(locales, options);
  }

  InitializeNumberFormat(this as any, locales, options, {
    getInternalSlots,
    localeData: NumberFormat.localeData,
    availableLocales: NumberFormat.availableLocales,
    getDefaultLocale: NumberFormat.getDefaultLocale,
    currencyDigitsData,
    numberingSystemNames,
  });

  const internalSlots = getInternalSlots(this as any);

  const dataLocale = internalSlots.dataLocale;
  const dataLocaleData = NumberFormat.localeData[dataLocale];
  invariant(
    dataLocaleData !== undefined,
    `Cannot load locale-dependent data for ${dataLocale}.`
  );

  internalSlots.pl = new Intl.PluralRules(dataLocale, {
    minimumFractionDigits: internalSlots.minimumFractionDigits,
    maximumFractionDigits: internalSlots.maximumFractionDigits,
    minimumIntegerDigits: internalSlots.minimumIntegerDigits,
    minimumSignificantDigits: internalSlots.minimumSignificantDigits,
    maximumSignificantDigits: internalSlots.maximumSignificantDigits,
  } as any);
  return this;
} as NumberFormatConstructor;

defineProperty(NumberFormat.prototype, 'formatToParts', {
  value: function formatToParts(x: number) {
    return FormatNumericToParts(this, toNumeric(x) as number, {
      getInternalSlots,
    });
  },
});

defineProperty(NumberFormat.prototype, 'resolvedOptions', {
  value: function resolvedOptions() {
    if (typeof this !== 'object' || !(this instanceof NumberFormat)) {
      throw TypeError(
        'Method Intl.NumberFormat.prototype.resolvedOptions called on incompatible receiver'
      );
    }
    const internalSlots = getInternalSlots(this as any);
    const ro: Record<string, unknown> = {};
    for (const key of RESOLVED_OPTIONS_KEYS) {
      const value = internalSlots[key];
      if (value !== undefined) {
        ro[key] = value;
      }
    }
    return ro as any;
  },
});

const formatDescriptor = {
  enumerable: false,
  configurable: true,
  get() {
    if (typeof this !== 'object' || !(this instanceof NumberFormat)) {
      throw TypeError(
        'Intl.NumberFormat format property accessor called on incompatible receiver'
      );
    }
    const internalSlots = getInternalSlots(this as any);
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const numberFormat = this;
    let boundFormat = internalSlots.boundFormat;
    if (boundFormat === undefined) {
      // https://tc39.es/proposal-unified-intl-numberformat/section11/numberformat_diff_out.html#sec-number-format-functions
      boundFormat = (value?: number | bigint) => {
        // TODO: check bigint
        const x = toNumeric(value) as number;
        return numberFormat
          .formatToParts(x)
          .map(x => x.value)
          .join('');
      };
      try {
        // https://github.com/tc39/test262/blob/master/test/intl402/NumberFormat/prototype/format/format-function-name.js
        Object.defineProperty(boundFormat, 'name', {
          configurable: true,
          enumerable: false,
          writable: false,
          value: '',
        });
      } catch (e) {
        // In older browser (e.g Chrome 36 like polyfill.io)
        // TypeError: Cannot redefine property: name
      }
      internalSlots.boundFormat = boundFormat;
    }
    return boundFormat;
  },
} as const;
try {
  // https://github.com/tc39/test262/blob/master/test/intl402/NumberFormat/prototype/format/name.js
  Object.defineProperty(formatDescriptor.get, 'name', {
    configurable: true,
    enumerable: false,
    writable: false,
    value: 'get format',
  });
} catch (e) {
  // In older browser (e.g Chrome 36 like polyfill.io)
  // TypeError: Cannot redefine property: name
}

Object.defineProperty(NumberFormat.prototype, 'format', formatDescriptor);

// Static properties
defineProperty(NumberFormat, 'supportedLocalesOf', {
  value: function supportedLocalesOf(
    locales: string | string[],
    options?: Pick<NumberFormatOptions, 'localeMatcher'>
  ) {
    return SupportedLocales(
      NumberFormat.availableLocales,
      CanonicalizeLocaleList(locales),
      options
    );
  },
});

NumberFormat.__addLocaleData = function __addLocaleData(
  ...data: RawNumberLocaleData[]
) {
  for (const datum of data) {
    const availableLocales: string[] = datum.availableLocales;
    for (const locale of availableLocales) {
      try {
        NumberFormat.localeData[locale] = unpackData(locale, datum);
      } catch (e) {
        // Ignore if we got no data
      }
    }
  }
  NumberFormat.availableLocales = Object.keys(NumberFormat.localeData);
  if (!NumberFormat.__defaultLocale) {
    NumberFormat.__defaultLocale = NumberFormat.availableLocales[0];
  }
};

NumberFormat.__defaultLocale = 'en';
NumberFormat.localeData = {};
NumberFormat.availableLocales = [];
NumberFormat.getDefaultLocale = () => {
  return NumberFormat.__defaultLocale;
};
NumberFormat.polyfilled = true;

function toNumeric(val: any) {
  if (typeof val === 'bigint') {
    return val;
  }
  return ToNumber(val);
}

try {
  // IE11 does not have Symbol
  if (typeof Symbol !== 'undefined') {
    Object.defineProperty(NumberFormat.prototype, Symbol.toStringTag, {
      configurable: true,
      enumerable: false,
      writable: false,
      value: 'Intl.NumberFormat',
    });
  }

  // https://github.com/tc39/test262/blob/master/test/intl402/NumberFormat/length.js
  Object.defineProperty(NumberFormat.prototype.constructor, 'length', {
    configurable: true,
    enumerable: false,
    writable: false,
    value: 0,
  });
  // https://github.com/tc39/test262/blob/master/test/intl402/NumberFormat/supportedLocalesOf/length.js
  Object.defineProperty(NumberFormat.supportedLocalesOf, 'length', {
    configurable: true,
    enumerable: false,
    writable: false,
    value: 1,
  });

  Object.defineProperty(NumberFormat, 'prototype', {
    configurable: false,
    enumerable: false,
    writable: false,
    value: NumberFormat.prototype,
  });
} catch (e) {
  // Meta fix so we're test262-compliant, not important
}
