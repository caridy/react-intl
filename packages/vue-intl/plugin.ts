import {createIntl, IntlConfig} from '@formatjs/intl'
import Vue from 'vue'

export const plugin: Vue.Plugin = {
  install: (app, options: IntlConfig) => {
    if (!options) {
      throw new Error('Missing `options` for vue-intl plugin')
    }
    const intl = createIntl(options)

    app.config.globalProperties.$intl = intl
    app.config.globalProperties.$formatMessage = intl.formatMessage
    app.config.globalProperties.$formatDate = intl.formatDate
    app.config.globalProperties.$formatTime = intl.formatTime
    app.config.globalProperties.$formatDateTimeRange = intl.formatDateTimeRange
    app.config.globalProperties.$formatRelativeTime = intl.formatRelativeTime
    app.config.globalProperties.$formatDisplayName = intl.formatDisplayName
    app.config.globalProperties.$formatNumber = intl.formatNumber

    app.provide('intl', intl)
  },
}
