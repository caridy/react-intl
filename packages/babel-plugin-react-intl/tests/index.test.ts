import {join, basename} from 'path';

import {transformFileSync} from '@babel/core';
import plugin, {OptionsSchema} from '../';

const TESTS: Record<string, OptionsSchema> = {
  additionalComponentNames: {
    additionalComponentNames: ['CustomMessage'],
  },
  ast: {
    ast: true,
    extractFromFormatMessageCall: true,
  },
  defineMessage: {},
  defineMessages: {},
  descriptionsAsObjects: {},
  empty: {},
  enforceDefaultMessage: {},
  enforceDescriptions: {},
  extractFromFormatMessageCall: {
    extractFromFormatMessageCall: true,
  },
  extractFromFormatMessageCallStateless: {
    extractFromFormatMessageCall: true,
  },
  formatMessageCall: {},
  FormattedMessage: {},
  idInterpolationPattern: {
    idInterpolationPattern: '[sha512:contenthash:hex:6]',
  },
  inline: {},
  moduleSourceName: {
    moduleSourceName: 'react-i18n',
  },
  overrideIdFn: {
    overrideIdFn: (
      id?: string,
      defaultMessage?: string,
      description?: string,
      filePath?: string
    ) => {
      const filename = basename(filePath!);
      return `${filename}.${id}.${
        defaultMessage!.length
      }.${typeof description}`;
    },
  },
  removeDefaultMessage: {
    removeDefaultMessage: true,
  },
};

describe('emit asserts for: ', () => {
  for (const [caseName, opts] of Object.entries(TESTS)) {
    it(caseName, function () {
      const filePath = join(__dirname, 'fixtures', caseName, 'actual.js');
      const {
        code,
        // @ts-ignore
        metadata: {'react-intl': data},
      } = transform(filePath, {
        pragma: '@react-intl',
        ...opts,
      });
      expect({
        data,
        code: code?.trim(),
      }).toMatchSnapshot();
    });
  }
});

test('extractSourceLocation', function () {
  const filePath = join(
    __dirname,
    'fixtures',
    'extractSourceLocation',
    'actual.js'
  );
  const {
    code,
    // @ts-ignore
    metadata: {'react-intl': data},
  } = transform(filePath, {
    pragma: '@react-intl',
    extractSourceLocation: true,
  });
  expect(code?.trim()).toMatchSnapshot();
  expect(data).toMatchSnapshot({
    messages: [
      {
        file: expect.any(String),
        start: expect.any(Object),
        end: expect.any(Object),
      },
    ],
  });
});

describe('errors', () => {
  it('Properly throws parse errors', () => {
    expect(() =>
      transform(join(__dirname, 'fixtures', 'icuSyntax', 'actual.js'))
    ).toThrow(/Expected .* but "\." found/);
  });
});

let cacheBust = 1;

function transform(
  filePath: string,
  options = {},
  {multiplePasses = false} = {}
) {
  function getPluginConfig() {
    return [plugin, options, Date.now() + '' + ++cacheBust];
  }

  return transformFileSync(filePath, {
    plugins: multiplePasses
      ? [
          'module:@babel/plugin-syntax-jsx',
          getPluginConfig(),
          getPluginConfig(),
        ]
      : ['module:@babel/plugin-syntax-jsx', getPluginConfig()],
  })!;
}
