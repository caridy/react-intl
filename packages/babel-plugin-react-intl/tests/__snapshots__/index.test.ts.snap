// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emit asserts for:  FormattedMessage 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { FormattedMessage } from 'react-intl';
export default class Foo extends Component {
  render() {
    return <FormattedMessage id=\\"foo.bar.baz\\" defaultMessage=\\"Hello World!\\" />;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message.",
        "id": "foo.bar.baz",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  additionalComponentNames 1`] = `
Object {
  "code": "import React, { Component } from 'react';

function CustomMessage() {}

export default class Foo extends Component {
  render() {
    return <CustomMessage id=\\"greeting-world\\" defaultMessage=\\"Hello World!\\" />;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": "Greeting to the world",
        "id": "greeting-world",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  ast 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { FormattedMessage, defineMessage, defineMessages } from 'react-intl';
defineMessage({
  \\"id\\": \\"defineMessage\\",
  \\"defaultMessage\\": [{
    \\"type\\": 0,
    \\"value\\": \\"this is a \\"
  }, {
    \\"type\\": 3,
    \\"style\\": \\"full\\",
    \\"value\\": \\"dt\\"
  }]
});
defineMessages({
  foo: {
    \\"id\\": \\"defineMessages1\\",
    \\"defaultMessage\\": [{
      \\"type\\": 0,
      \\"value\\": \\"this is a \\"
    }, {
      \\"type\\": 4,
      \\"style\\": \\"full\\",
      \\"value\\": \\"dt\\"
    }]
  },
  bar: {
    \\"id\\": \\"defineMessages2\\",
    \\"defaultMessage\\": [{
      \\"type\\": 0,
      \\"value\\": \\"this is a \\"
    }, {
      \\"type\\": 2,
      \\"style\\": null,
      \\"value\\": \\"dt\\"
    }]
  }
});
export default class Foo extends Component {
  render() {
    Intl.formatMessage({
      id: 'intl.formatMessage',
      defaultMessage: 'foo {s, plural, one{1} other{2}}'
    });
    return <FormattedMessage id=\\"foo.bar.baz\\" defaultMessage={[{
      \\"type\\": 0,
      \\"value\\": \\"Hello World!\\"
    }]} />;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "this is a {dt, date, full}",
        "description": undefined,
        "id": "defineMessage",
      },
      Object {
        "defaultMessage": "this is a {dt, time, full}",
        "description": undefined,
        "id": "defineMessages1",
      },
      Object {
        "defaultMessage": "this is a {dt, number}",
        "description": undefined,
        "id": "defineMessages2",
      },
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message.",
        "id": "foo.bar.baz",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  defineMessage 1`] = `
Object {
  "code": "// @react-intl project:amazing
import React, { Component } from 'react';
import { defineMessage, FormattedMessage } from 'react-intl';
const msgs = {
  header: defineMessage({
    \\"id\\": \\"foo.bar.baz\\",
    \\"defaultMessage\\": \\"Hello World!\\"
  }),
  content: defineMessage({
    \\"id\\": \\"foo.bar.biff\\",
    \\"defaultMessage\\": \\"Hello Nurse!\\"
  }),
  kittens: defineMessage({
    \\"id\\": \\"app.home.kittens\\",
    \\"defaultMessage\\": \\"{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}\\"
  }),
  trailingWhitespace: defineMessage({
    \\"id\\": \\"trailing.ws\\",
    \\"defaultMessage\\": \\"   Some whitespace   \\"
  }),
  escaped: defineMessage({
    \\"id\\": \\"escaped.apostrophe\\",
    \\"defaultMessage\\": \\"A quoted value ''{value}'\\"
  })
};
export default class Foo extends Component {
  render() {
    return <div>
        <h1>
          <FormattedMessage {...msgs.header} />
        </h1>
        <p>
          <FormattedMessage {...msgs.content} />
        </p>
        <p>
          <FormattedMessage {...msgs.kittens} />
        </p>
      </div>;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message",
        "id": "foo.bar.baz",
      },
      Object {
        "defaultMessage": "Hello Nurse!",
        "description": "Another message",
        "id": "foo.bar.biff",
      },
      Object {
        "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
        "description": "Counts kittens",
        "id": "app.home.kittens",
      },
      Object {
        "defaultMessage": "   Some whitespace   ",
        "description": "Whitespace",
        "id": "trailing.ws",
      },
      Object {
        "defaultMessage": "A quoted value ''{value}'",
        "description": "Escaped apostrophe",
        "id": "escaped.apostrophe",
      },
    ],
    "meta": Object {
      "project": "amazing",
    },
  },
}
`;

exports[`emit asserts for:  defineMessages 1`] = `
Object {
  "code": "// @react-intl project:amazing
import React, { Component } from 'react';
import { defineMessages, FormattedMessage } from 'react-intl';
const msgs = defineMessages({
  header: {
    \\"id\\": \\"foo.bar.baz\\",
    \\"defaultMessage\\": \\"Hello World!\\"
  },
  content: {
    \\"id\\": \\"foo.bar.biff\\",
    \\"defaultMessage\\": \\"Hello Nurse!\\"
  },
  kittens: {
    \\"id\\": \\"app.home.kittens\\",
    \\"defaultMessage\\": \\"{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}\\"
  },
  trailingWhitespace: {
    \\"id\\": \\"trailing.ws\\",
    \\"defaultMessage\\": \\"   Some whitespace   \\"
  },
  escaped: {
    \\"id\\": \\"escaped.apostrophe\\",
    \\"defaultMessage\\": \\"A quoted value ''{value}'\\"
  }
});
export default class Foo extends Component {
  render() {
    return <div>
        <h1>
          <FormattedMessage {...msgs.header} />
        </h1>
        <p>
          <FormattedMessage {...msgs.content} />
        </p>
        <p>
          <FormattedMessage {...msgs.kittens} />
        </p>
      </div>;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message",
        "id": "foo.bar.baz",
      },
      Object {
        "defaultMessage": "Hello Nurse!",
        "description": "Another message",
        "id": "foo.bar.biff",
      },
      Object {
        "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
        "description": "Counts kittens",
        "id": "app.home.kittens",
      },
      Object {
        "defaultMessage": "   Some whitespace   ",
        "description": "Whitespace",
        "id": "trailing.ws",
      },
      Object {
        "defaultMessage": "A quoted value ''{value}'",
        "description": "Escaped apostrophe",
        "id": "escaped.apostrophe",
      },
    ],
    "meta": Object {
      "project": "amazing",
    },
  },
}
`;

exports[`emit asserts for:  descriptionsAsObjects 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { FormattedMessage } from 'react-intl'; // @react-intl project:amazing2

export default class Foo extends Component {
  render() {
    return <FormattedMessage id=\\"foo.bar.baz\\" defaultMessage=\\"Hello World!\\" />;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": Object {
          "metadata": "Additional metadata content.",
          "text": "Something for the translator.",
        },
        "id": "foo.bar.baz",
      },
    ],
    "meta": Object {
      "project": "amazing2",
    },
  },
}
`;

exports[`emit asserts for:  empty 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { defineMessage } from 'react-intl';
export default class Foo extends Component {
  render() {
    return <div></div>;
  }

}",
  "data": Object {
    "messages": Array [],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  enforceDefaultMessage 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { FormattedMessage, defineMessages } from 'react-intl';
defineMessages({
  test: {
    \\"id\\": \\"TRANSLATION_KEY\\"
  }
});
const testMessage = {
  id: 'TRANSLATION_KEY2'
};
export default class Foo extends Component {
  render() {
    return <>
        <FormattedMessage id=\\"foo.bar.baz\\" />
        <FormattedMessage {...testMessage} />
      </>;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": undefined,
        "description": undefined,
        "id": "TRANSLATION_KEY",
      },
      Object {
        "defaultMessage": undefined,
        "description": undefined,
        "id": "foo.bar.baz",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  enforceDescriptions 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { FormattedMessage } from 'react-intl';
export default class Foo extends Component {
  render() {
    return <FormattedMessage id=\\"foo.bar.baz\\" defaultMessage=\\"Hello World!\\" />;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": undefined,
        "id": "foo.bar.baz",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  extractFromFormatMessageCall 1`] = `
Object {
  "code": "import { FormattedMessage, injectIntl } from 'react-intl';
import React, { Component } from 'react';
const objectPointer = {
  id: 'foo.bar.invalid',
  defaultMessage: 'This cannot be extracted',
  description: 'the plugin only supports inline objects'
};

class Foo extends Component {
  render() {
    const {
      intl
    } = this.props;
    const {
      intl: {
        formatMessage
      }
    } = this.props;
    const msgs = {
      baz: this.props.intl.formatMessage({
        \\"id\\": \\"foo.bar.baz\\",
        \\"defaultMessage\\": \\"Hello World!\\"
      }),
      biff: intl.formatMessage({
        \\"id\\": \\"foo.bar.biff\\",
        \\"defaultMessage\\": \\"Hello Nurse!\\"
      }),
      qux: formatMessage({
        \\"id\\": \\"foo.bar.qux\\",
        \\"defaultMessage\\": \\"Hello Stranger!\\"
      }),
      invalid: this.props.intl.formatMessage(objectPointer)
    };
    return <div>
        <h1>{msgs.header}</h1>
        <p>{msgs.content}</p>
        <span>
          <FormattedMessage id=\\"foo\\" defaultMessage=\\"bar\\" />
        </span>
      </div>;
  }

}

export default injectIntl(Foo);",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message",
        "id": "foo.bar.baz",
      },
      Object {
        "defaultMessage": "Hello Nurse!",
        "description": "Another message",
        "id": "foo.bar.biff",
      },
      Object {
        "defaultMessage": "Hello Stranger!",
        "description": "A different message",
        "id": "foo.bar.qux",
      },
      Object {
        "defaultMessage": "bar",
        "description": "baz",
        "id": "foo",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  extractFromFormatMessageCallStateless 1`] = `
Object {
  "code": "import { FormattedMessage, injectIntl, useIntl } from 'react-intl';
import React from 'react';

function myFunction(param1, {
  formatMessage,
  formatDate
}) {
  return formatMessage({
    id: 'inline1',
    defaultMessage: 'Hello params!',
    description: 'A stateless message'
  }) + formatDate(new Date());
}

const child = myFunction(filterable, intl);

function SFC() {
  const {
    formatMessage
  } = useIntl();
  return formatMessage({
    \\"id\\": \\"hook\\",
    \\"defaultMessage\\": \\"hook <b>foo</b>\\"
  });
}

const Foo = ({
  intl: {
    formatMessage
  }
}) => {
  const msgs = {
    qux: formatMessage({
      id: 'foo.bar.quux',
      defaultMessage: 'Hello <b>Stateless!</b>',
      description: 'A stateless message'
    })
  };
  return <div>
      <h1>{msgs.header}</h1>
      <p>{msgs.content}</p>
      <span>
        <FormattedMessage id=\\"foo\\" defaultMessage=\\"bar\\" />
      </span>
    </div>;
};

export default injectIntl(Foo);",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "hook <b>foo</b>",
        "description": "hook",
        "id": "hook",
      },
      Object {
        "defaultMessage": "bar",
        "description": "baz",
        "id": "foo",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  formatMessageCall 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { injectIntl, FormattedMessage } from 'react-intl';
const objectPointer = {
  id: 'foo.bar.invalid',
  defaultMessage: 'This cannot be extracted',
  description: 'the plugin only supports inline objects'
};

class Foo extends Component {
  render() {
    const msgs = {
      baz: this.props.intl.formatMessage({
        id: 'foo.bar.baz',
        defaultMessage: 'Hello World!',
        description: 'The default message'
      }),
      biff: this.props.intl.formatMessage({
        id: 'foo.bar.biff',
        defaultMessage: 'Hello Nurse!',
        description: 'Another message'
      }),
      invalid: this.props.intl.formatMessage(objectPointer)
    };
    return <div>
        <h1>{msgs.header}</h1>
        <p>{msgs.content}</p>
        <span>
          <FormattedMessage id=\\"foo\\" defaultMessage=\\"bar\\" />
        </span>
      </div>;
  }

}

export default injectIntl(Foo);",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "bar",
        "description": "baz",
        "id": "foo",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  idInterpolationPattern 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { defineMessages, FormattedMessage } from 'react-intl';
const msgs = defineMessages({
  header: {
    \\"id\\": \\"4d1460\\",
    \\"defaultMessage\\": \\"Hello World!\\"
  },
  content: {
    \\"id\\": \\"foo.bar.biff\\",
    \\"defaultMessage\\": \\"Hello Nurse!\\"
  }
});
export default class Foo extends Component {
  render() {
    return <div>
        <h1>
          <FormattedMessage {...msgs.header} />
        </h1>
        <p>
          <FormattedMessage {...msgs.content} />
        </p>
        <FormattedMessage id=\\"36a8c8\\" defaultMessage=\\"Hello World!\\" />
        <FormattedMessage id=\\"5ce864\\" defaultMessage=\\"NO ID\\" />
      </div>;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message",
        "id": "4d1460",
      },
      Object {
        "defaultMessage": "Hello Nurse!",
        "description": Object {
          "metadata": "Additional metadata content.",
          "text": "Something for the translator.",
        },
        "id": "foo.bar.biff",
      },
      Object {
        "defaultMessage": "Hello World!",
        "description": "Something for the translator. Another description",
        "id": "36a8c8",
      },
      Object {
        "defaultMessage": "NO ID",
        "description": "Something for the translator. Another description",
        "id": "5ce864",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  inline 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { FormattedMessage, defineMessage } from 'react-intl';
export default class Foo extends Component {
  render() {
    return <div>
        <FormattedMessage id=\\"foo.bar.baz\\" defaultMessage=\\"Hello World!\\" />
        {defineMessage({
        \\"id\\": \\"header\\",
        \\"defaultMessage\\": \\"Hello World!\\"
      })}
        {defineMessage({
        \\"id\\": \\"header2\\",
        \\"defaultMessage\\": \\"Hello World!\\"
      })}
      </div>;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message.",
        "id": "foo.bar.baz",
      },
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message",
        "id": "header",
      },
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message",
        "id": "header2",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  moduleSourceName 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { FormattedMessage } from 'react-i18n';
import { defineMessages } from 'react-intl'; // These should be ignored because the \`moduleSourceName\` is configured to be:
// \`react-i18n\`.

const msgs = defineMessages({
  header: {
    id: 'header',
    defaultMessage: 'Hello World!',
    description: 'The default message'
  },
  content: {
    id: 'content',
    defaultMessage: 'Hello Nurse!',
    description: 'Another message'
  }
});
export default class Foo extends Component {
  render() {
    return <div>
        <FormattedMessage id=\\"foo.bar.baz\\" defaultMessage=\\"Hello World!\\" />
        {msgs}
      </div>;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message.",
        "id": "foo.bar.baz",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  overrideIdFn 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { defineMessages, FormattedMessage } from 'react-intl';
const msgs = defineMessages({
  header: {
    \\"id\\": \\"actual.js.foo.bar.baz.12.string\\",
    \\"defaultMessage\\": \\"Hello World!\\"
  },
  content: {
    \\"id\\": \\"actual.js.foo.bar.biff.12.object\\",
    \\"defaultMessage\\": \\"Hello Nurse!\\"
  }
});
export default class Foo extends Component {
  render() {
    return <div>
        <h1>
          <FormattedMessage {...msgs.header} />
        </h1>
        <p>
          <FormattedMessage {...msgs.content} />
        </p>
        <FormattedMessage id=\\"actual.js.foo.bar.zoo.12.object\\" defaultMessage=\\"Hello World!\\" />
        <FormattedMessage id=\\"actual.js..5.object\\" defaultMessage=\\"NO ID\\" />
      </div>;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello World!",
        "description": "The default message",
        "id": "actual.js.foo.bar.baz.12.string",
      },
      Object {
        "defaultMessage": "Hello Nurse!",
        "description": Object {
          "metadata": "Additional metadata content.",
          "text": "Something for the translator.",
        },
        "id": "actual.js.foo.bar.biff.12.object",
      },
      Object {
        "defaultMessage": "Hello World!",
        "description": Object {
          "metadata": "Additional metadata content.",
          "text": "Something for the translator. Another description",
        },
        "id": "actual.js.foo.bar.zoo.12.object",
      },
      Object {
        "defaultMessage": "NO ID",
        "description": Object {
          "metadata": "Additional metadata content.",
          "text": "Something for the translator. Another description",
        },
        "id": "actual.js..5.object",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`emit asserts for:  removeDefaultMessage 1`] = `
Object {
  "code": "import React, { Component } from 'react';
import { defineMessages, FormattedMessage } from 'react-intl';
const messages = defineMessages({
  foo: {
    \\"id\\": \\"greeting-user\\"
  }
});
export default class Foo extends Component {
  render() {
    return <FormattedMessage id=\\"greeting-world\\" />;
  }

}",
  "data": Object {
    "messages": Array [
      Object {
        "defaultMessage": "Hello, {name}",
        "description": "Greeting the user",
        "id": "greeting-user",
      },
      Object {
        "defaultMessage": "Hello World!",
        "description": "Greeting to the world",
        "id": "greeting-world",
      },
    ],
    "meta": Object {},
  },
}
`;

exports[`extractSourceLocation 1`] = `
"import React, { Component } from 'react';
import { FormattedMessage } from 'react-intl';
export default class Foo extends Component {
  render() {
    return <FormattedMessage id=\\"foo.bar.baz\\" defaultMessage=\\"Hello World!\\" />;
  }

}"
`;

exports[`extractSourceLocation 2`] = `
Object {
  "messages": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": undefined,
      "end": Any<Object>,
      "file": Any<String>,
      "id": "foo.bar.baz",
      "start": Any<Object>,
    },
  ],
  "meta": Object {},
}
`;
